baseapp/baseapp.go

```diff
@@ -540,17 +536,60 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 		ctx = ctx.WithMultiStore(msCache)
 	}
 
-	result = handler(ctx, msg)
+	finalResult := sdk.Result{}
+	var logs []string
+	for i, msg := range msgs {
+		// Match route.
+		msgType := msg.Type()
+		handler := app.router.Route(msgType)
+		if handler == nil {
+			return sdk.ErrUnknownRequest("Unrecognized Msg type: " + msgType).Result()
+		}
+
+		result = handler(ctx, msg)
+
+		// Set gas utilized
+		finalResult.GasUsed += ctx.GasMeter().GasConsumed()
+		finalResult.GasWanted += result.GasWanted
+
+		// Append Data and Tags
+		finalResult.Data = append(finalResult.Data, result.Data...)
+		finalResult.Tags = append(finalResult.Tags, result.Tags...)
 
-	// Set gas utilized
-	result.GasUsed = ctx.GasMeter().GasConsumed()
+		// Construct usable logs in multi-message transactions. Messages are 1-indexed in logs.
+		logs = append(logs, fmt.Sprintf("Msg %d: %s", i + 1, finalResult.Log))
+
+		// Cumulate Validator updates
+		for _, val := range result.ValidatorUpdates {
```

Comment:
- I think this logic isn't quite right. Need to deal with the case of a power *decrease* too (which will have a lower power, not a negative power). Look at what Tendermint does to compute the change.
- Can we separate this out into a function and add a few unit tests? Clearer & might be useful elsewhere.

Reply:
Haha, nevermind - `ValidatorSetUpdates` should never have been in the result type to begin with.

Reply:
So deliverTx should never change the validator set, right? It all happens in EndBlocker?